### ISSUE TITLE
link errors in vcpkg

### ISSUE THREAD

**Issue by pmahend1** [arvidn/libtorrent#3878]
**Please provide the following information**

libtorrent version (or branch): vcpkg  2019-04-19

platform/architecture: windows 64

compiler and compiler version: msvc

please describe what symptom you see, what you would expect to see instead and
how to reproduce it.

I tried to run demo example I am getting the following error on compile
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK2019	unresolved external symbol "public: __cdecl libtorrent::internal_file_entry::~internal_file_entry(void)" (??1internal_file_entry@libtorrent@@QEAA@XZ) referenced in function "public: void * __cdecl libtorrent::internal_file_entry::`scalar deleting destructor'(unsigned int)" (??_Ginternal_file_entry@libtorrent@@QEAAPEAXI@Z)	libtorrent	C:\workspace\MyProject-POC\libtorrent\libtorrent_client.obj	1	
Error	LNK2001	unresolved external symbol "public: __cdecl libtorrent::internal_file_entry::~internal_file_entry(void)" (??1internal_file_entry@libtorrent@@QEAA@XZ)	libtorrent	C:\workspace\MyProject-POC\libtorrent\Main.obj	1	
Error	LNK2019	unresolved external symbol "public: __cdecl libtorrent::add_torrent_params::add_torrent_params(class std::function<struct libtorrent::storage_interface * __cdecl(struct libtorrent::storage_params const &,struct libtorrent::file_pool &)>)" (??0add_torrent_params@libtorrent@@QEAA@V?$function@$$A6APEAUstorage_interface@libtorrent@@AEBUstorage_params@2@AEAUfile_pool@2@@Z@std@@@Z) referenced in function "public: void __cdecl libtorrent::add_torrent_params::`default constructor closure'(void)" (??_Fadd_torrent_params@libtorrent@@QEAAXXZ)	libtorrent	C:\workspace\MyProject-POC\libtorrent\libtorrent_client.obj	1	
Error	LNK2001	unresolved external symbol "public: __cdecl libtorrent::add_torrent_params::add_torrent_params(class std::function<struct libtorrent::storage_interface * __cdecl(struct libtorrent::storage_params const &,struct libtorrent::file_pool &)>)" (??0add_torrent_params@libtorrent@@QEAA@V?$function@$$A6APEAUstorage_interface@libtorrent@@AEBUstorage_params@2@AEAUfile_pool@2@@Z@std@@@Z)	libtorrent	C:\workspace\MyProject-POC\libtorrent\Main.obj	1	
Error	LNK2019	unresolved external symbol "public: struct libtorrent::torrent_handle __cdecl libtorrent::session_handle::add_torrent(struct libtorrent::add_torrent_params const &)" (?add_torrent@session_handle@libtorrent@@QEAA?AUtorrent_handle@2@AEBUadd_torrent_params@2@@Z) referenced in function "public: void __cdecl Client::StartClientForFile(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?StartClientForFile@Client@@QEAAXAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)	libtorrent	C:\workspace\MyProject-POC\libtorrent\libtorrent_client.obj	1	
Error	LNK2019	unresolved external symbol "public: struct libtorrent::file_entry __cdecl libtorrent::file_storage::at_deprecated(int)const " (?at_deprecated@file_storage@libtorrent@@QEBA?AUfile_entry@2@H@Z) referenced in function "public: struct libtorrent::file_entry __cdecl libtorrent::torrent_info::file_at(int)const " (?file_at@torrent_info@libtorrent@@QEBA?AUfile_entry@2@H@Z)	libtorrent	C:\workspace\MyProject-POC\libtorrent\libtorrent_client.obj	1	
Error	LNK2001	unresolved external symbol "public: struct libtorrent::file_entry __cdecl libtorrent::file_storage::at_deprecated(int)const " (?at_deprecated@file_storage@libtorrent@@QEBA?AUfile_entry@2@H@Z)	libtorrent	C:\workspace\MyProject-POC\libtorrent\Main.obj	1	
Error	LNK2019	unresolved external symbol "public: class std::_Vector_const_iterator<class std::_Vector_val<struct std::_Simple_types<struct libtorrent::internal_file_entry> > > __cdecl libtorrent::file_storage::file_at_offset_deprecated(__int64)const " (?file_at_offset_deprecated@file_storage@libtorrent@@QEBA?AV?$_Vector_const_iterator@V?$_Vector_val@U?$_Simple_types@Uinternal_file_entry@libtorrent@@@std@@@std@@@std@@_J@Z) referenced in function "public: class std::_Vector_const_iterator<class std::_Vector_val<struct std::_Simple_types<struct libtorrent::internal_file_entry> > > __cdecl libtorrent::torrent_info::file_at_offset(__int64)const " (?file_at_offset@torrent_info@libtorrent@@QEBA?AV?$_Vector_const_iterator@V?$_Vector_val@U?$_Simple_types@Uinternal_file_entry@libtorrent@@@std@@@std@@@std@@_J@Z)	libtorrent	C:\workspace\MyProject-POC\libtorrent\libtorrent_client.obj	1	
Error	LNK2001	unresolved external symbol "public: class std::_Vector_const_iterator<class std::_Vector_val<struct std::_Simple_types<struct libtorrent::internal_file_entry> > > __cdecl libtorrent::file_storage::file_at_offset_deprecated(__int64)const " (?file_at_offset_deprecated@file_storage@libtorrent@@QEBA?AV?$_Vector_const_iterator@V?$_Vector_val@U?$_Simple_types@Uinternal_file_entry@libtorrent@@@std@@@std@@@std@@_J@Z)	libtorrent	C:\workspace\MyProject-POC\libtorrent\Main.obj	1	
Error	LNK2019	unresolved external symbol "public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl libtorrent::fingerprint::to_string(void)const " (?to_string@fingerprint@libtorrent@@QEBA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ) referenced in function "public: __cdecl libtorrent::session::session(struct libtorrent::fingerprint const &,struct libtorrent::flags::bitfield_flag<unsigned char,struct libtorrent::session_flags_tag,void>,struct libtorrent::flags::bitfield_flag<unsigned int,struct libtorrent::alert_category_tag,void>)" (??0session@libtorrent@@QEAA@AEBUfingerprint@1@U?$bitfield_flag@EUsession_flags_tag@libtorrent@@X@flags@1@U?$bitfield_flag@IUalert_category_tag@libtorrent@@X@41@@Z)	libtorrent	C:\workspace\MyProject-POC\libtorrent\libtorrent_client.obj	1	
Error	LNK2001	unresolved external symbol "public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl libtorrent::fingerprint::to_string(void)const " (?to_string@fingerprint@libtorrent@@QEBA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ)	libtorrent	C:\workspace\MyProject-POC\libtorrent\Main.obj	1	
Error	LNK2001	unresolved external symbol "class libtorrent::from_span_t libtorrent::from_span" (?from_span@libtorrent@@3Vfrom_span_t@1@A)	libtorrent	C:\workspace\MyProject-POC\libtorrent\libtorrent_client.obj	1	
Error	LNK2001	unresolved external symbol "class libtorrent::from_span_t libtorrent::from_span" (?from_span@libtorrent@@3Vfrom_span_t@1@A)	libtorrent	C:\workspace\MyProject-POC\libtorrent\Main.obj	1	
Error	LNK1120	7 unresolved externals	libtorrent	C:\workspace\MyProject-POC\x64\Debug\libtorrent.exe	1	

```

Debug

----
```
1>------ Build started: Project: libtorrent, Configuration: Debug x64 ------
1>Main.cpp
1>C:\workspace\vcpkg\installed\x64-windows\include\libtorrent\file_pool.hpp(57,2): warning C4275:  non dll-interface class 'boost::noncopyable_::noncopyable' used as base for dll-interface struct 'libtorrent::file_pool'
1>C:\workspace\vcpkg\installed\x64-windows\include\boost\core\noncopyable.hpp(38): message :  see declaration of 'boost::noncopyable_::noncopyable'
1>libtorrent_client.cpp
1>Generating Code...
1>C:\workspace\vcpkg\installed\x64-windows\include\libtorrent\file_pool.hpp(56): message :  see declaration of 'libtorrent::file_pool'
1>C:\workspace\vcpkg\installed\x64-windows\include\libtorrent\file_pool.hpp(56): message : Please define _WIN32_WINNT or _WIN32_WINDOWS appropriately. For example:
1>C:\workspace\vcpkg\installed\x64-windows\include\libtorrent\file_pool.hpp(56): message : - add -D_WIN32_WINNT=0x0601 to the compiler command line; or
1>C:\workspace\vcpkg\installed\x64-windows\include\libtorrent\file_pool.hpp(56): message : - add _WIN32_WINNT=0x0601 to your project's Preprocessor Definitions.
1>C:\workspace\vcpkg\installed\x64-windows\include\libtorrent\file_pool.hpp(56): message : Assuming _WIN32_WINNT=0x0601 (i.e. Windows 7 target).
1>   Creating library C:\workspace\MyProject-POC\x64\Debug\libtorrent.lib and object C:\workspace\MyProject-POC\x64\Debug\libtorrent.exp
1>libtorrent_client.obj : error LNK2019: unresolved external symbol "public: __cdecl libtorrent::add_torrent_params::add_torrent_params(class std::function<struct libtorrent::storage_interface * __cdecl(struct libtorrent::storage_params const &,struct libtorrent::file_pool &)>)" (??0add_torrent_params@libtorrent@@QEAA@V?$function@$$A6APEAUstorage_interface@libtorrent@@AEBUstorage_params@2@AEAUfile_pool@2@@Z@std@@@Z) referenced in function "public: void __cdecl libtorrent::add_torrent_params::`default constructor closure'(void)" (??_Fadd_torrent_params@libtorrent@@QEAAXXZ)
1>Main.obj : error LNK2001: unresolved external symbol "public: __cdecl libtorrent::add_torrent_params::add_torrent_params(class std::function<struct libtorrent::storage_interface * __cdecl(struct libtorrent::storage_params const &,struct libtorrent::file_pool &)>)" (??0add_torrent_params@libtorrent@@QEAA@V?$function@$$A6APEAUstorage_interface@libtorrent@@AEBUstorage_params@2@AEAUfile_pool@2@@Z@std@@@Z)
1>libtorrent_client.obj : error LNK2019: unresolved external symbol "public: __cdecl libtorrent::internal_file_entry::~internal_file_entry(void)" (??1internal_file_entry@libtorrent@@QEAA@XZ) referenced in function "public: void * __cdecl libtorrent::internal_file_entry::`scalar deleting destructor'(unsigned int)" (??_Ginternal_file_entry@libtorrent@@QEAAPEAXI@Z)
1>Main.obj : error LNK2001: unresolved external symbol "public: __cdecl libtorrent::internal_file_entry::~internal_file_entry(void)" (??1internal_file_entry@libtorrent@@QEAA@XZ)
1>libtorrent_client.obj : error LNK2019: unresolved external symbol "public: class std::_Vector_const_iterator<class std::_Vector_val<struct std::_Simple_types<struct libtorrent::internal_file_entry> > > __cdecl libtorrent::file_storage::file_at_offset_deprecated(__int64)const " (?file_at_offset_deprecated@file_storage@libtorrent@@QEBA?AV?$_Vector_const_iterator@V?$_Vector_val@U?$_Simple_types@Uinternal_file_entry@libtorrent@@@std@@@std@@@std@@_J@Z) referenced in function "public: class std::_Vector_const_iterator<class std::_Vector_val<struct std::_Simple_types<struct libtorrent::internal_file_entry> > > __cdecl libtorrent::torrent_info::file_at_offset(__int64)const " (?file_at_offset@torrent_info@libtorrent@@QEBA?AV?$_Vector_const_iterator@V?$_Vector_val@U?$_Simple_types@Uinternal_file_entry@libtorrent@@@std@@@std@@@std@@_J@Z)
1>Main.obj : error LNK2001: unresolved external symbol "public: class std::_Vector_const_iterator<class std::_Vector_val<struct std::_Simple_types<struct libtorrent::internal_file_entry> > > __cdecl libtorrent::file_storage::file_at_offset_deprecated(__int64)const " (?file_at_offset_deprecated@file_storage@libtorrent@@QEBA?AV?$_Vector_const_iterator@V?$_Vector_val@U?$_Simple_types@Uinternal_file_entry@libtorrent@@@std@@@std@@@std@@_J@Z)
1>libtorrent_client.obj : error LNK2019: unresolved external symbol "public: struct libtorrent::file_entry __cdecl libtorrent::file_storage::at_deprecated(int)const " (?at_deprecated@file_storage@libtorrent@@QEBA?AUfile_entry@2@H@Z) referenced in function "public: struct libtorrent::file_entry __cdecl libtorrent::torrent_info::file_at(int)const " (?file_at@torrent_info@libtorrent@@QEBA?AUfile_entry@2@H@Z)
1>Main.obj : error LNK2001: unresolved external symbol "public: struct libtorrent::file_entry __cdecl libtorrent::file_storage::at_deprecated(int)const " (?at_deprecated@file_storage@libtorrent@@QEBA?AUfile_entry@2@H@Z)
1>libtorrent_client.obj : error LNK2019: unresolved external symbol "public: struct libtorrent::torrent_handle __cdecl libtorrent::session_handle::add_torrent(struct libtorrent::add_torrent_params const &)" (?add_torrent@session_handle@libtorrent@@QEAA?AUtorrent_handle@2@AEBUadd_torrent_params@2@@Z) referenced in function "public: void __cdecl Client::StartClientForFile(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?StartClientForFile@Client@@QEAAXAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
1>libtorrent_client.obj : error LNK2019: unresolved external symbol "public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl libtorrent::fingerprint::to_string(void)const " (?to_string@fingerprint@libtorrent@@QEBA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ) referenced in function "public: __cdecl libtorrent::session::session(struct libtorrent::fingerprint const &,struct libtorrent::flags::bitfield_flag<unsigned char,struct libtorrent::session_flags_tag,void>,struct libtorrent::flags::bitfield_flag<unsigned int,struct libtorrent::alert_category_tag,void>)" (??0session@libtorrent@@QEAA@AEBUfingerprint@1@U?$bitfield_flag@EUsession_flags_tag@libtorrent@@X@flags@1@U?$bitfield_flag@IUalert_category_tag@libtorrent@@X@41@@Z)
1>Main.obj : error LNK2001: unresolved external symbol "public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl libtorrent::fingerprint::to_string(void)const " (?to_string@fingerprint@libtorrent@@QEBA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ)
1>libtorrent_client.obj : error LNK2001: unresolved external symbol "class libtorrent::from_span_t libtorrent::from_span" (?from_span@libtorrent@@3Vfrom_span_t@1@A)
1>Main.obj : error LNK2001: unresolved external symbol "class libtorrent::from_span_t libtorrent::from_span" (?from_span@libtorrent@@3Vfrom_span_t@1@A)
1>C:\workspace\MyProject-POC\x64\Debug\libtorrent.exe : fatal error LNK1120: 7 unresolved externals
1>Done building project "libtorrent.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

**Comment by arvidn** [comment]
make sure you define the same `TORRENT_*` defines in the client as the library is built for. specifically, perhaps libtorrent is build with `TORRENT_NO_DEPRECATED` defined, but maybe not the client.
could you find out which version of libtorrent you're using? look in `include/libtorrent/version.hpp`
how do you link against libtorrent?

**Comment by pmahend1** [comment]
Do you mean ,I have to add in the client all the macros in entire library , like below?

``` c++
#ifdef TORRENT_WINDOWS
#include <direct.h> // for _getcwd
#endif

```

**Comment by arvidn** [comment]
no, not all the include guards, just the configuration macros. you could try adding `TORRENT_NO_DEPRECATE` to your project file.
how are you linking against `libtorrent.lib`?

**Comment by pmahend1** [comment]
Will check that. You mean libtorrent-faster?

Is there another lib file that I am unaware of which I should add lib dependency through project linker settings?

**Comment by arvidn** [comment]
The output you posted doesn't print the actual command line for the linker, which could be useful to ensure that it's actually linking against libtorrent, that's all.

I would expect vcpkg to include a feature where build features propagate to *dependees*, i.e. where libtorrent.lib makes your build automatically add it's include file directory and also build configurations like `TORRENT_NO_DEPRECATE`. Perhaps it's misconfigured to not do that, or perhaps you're expected to do that yourself.

**Comment by arvidn** [comment]
it's still unclear though. `client_test.cpp` shouldn't rely on any deprecated features. you may want to use `objdump` (or whatever msvc's tool is called) to ensure the library file actually has those symbols.

it's actually not clear to me what's failing. is `libtorrent.lib` failing to be created?

**Comment by ras0219-msft** [comment]
I think I've found the issue; @pmahend1, could you try the fixes in https://github.com/microsoft/vcpkg/pull/7345 and see if it resolves your issue? Otherwise, could we continue the discussion in that PR until we're sure it isn't a vcpkg-specific issue?

**Comment by arvidn** [comment]
that makes sense. if `TORRENT_LINKING_SHARED` isn't set on the client when linking against libtorrent as a shared library, then the headers won't have the required `declspec(dllimport)` annotatoins.

**Comment by pmahend1** [comment]
@ras0219-msft I made those changes . What should I do after that? Reinstall libtorrent?

**Comment by pmahend1** [comment]
> 
> 
> that makes sense. if `TORRENT_LINKING_SHARED` isn't set on the client when linking against libtorrent as a shared library, then the headers won't have the required `declspec(dllimport)` annotatoins.

Do I have to add below line in client cpp file?
```
#ifndef TORRENT_LINKING_SHARED
#define TORRENT_LINKING_SHARED
#endif
```

**Comment by arvidn** [comment]
@pmahend1 you have to add the `TORRENT_LINKING_SHARED` define to your project.
If I read the patch in https://github.com/microsoft/vcpkg/pull/7345 correctly, it's actually updating the `libtorrent/export.hpp` header to always define this (presumably after it has built the library itself).

**Comment by pmahend1** [comment]
@arvidn I couldnt find export.hpp file in the include folder. I also compiled from source by referring https://libtorrent.org/building.html

If I compile with win32 it says its not able to file libkernel32.lib and with 64 bit I get same linking error.

**Comment by arvidn** [comment]
export.hpp is here: https://github.com/arvidn/libtorrent/blob/RC_1_2/include/libtorrent/aux_/export.hpp

Yes, that's where you add defines. but, as we've established by the vcpkg patch, the define you need to add is `TORRENT_LINKING_SHARED`

**Comment by pmahend1** [comment]
I added `TORRENT_LINKING_SHARED` and `TORRENT_NO_DEPRECATE` but still get linking issues. I will check more. BTW it was `TORRENT_NO_DEPRECATE` , not` TORRENT_NO_DEPRECATED`

**Comment by pmahend1** [comment]
Somehow I got rid off all the linking issues. Not sure which one worked.
- I edited port file as mentioned by @ras0219-msft 
- rebuilt libtorrent by `vcpkg build libtorrent:<triplet>` and then started fresh project.
- Only copied torrent client from example.
- added preprocessor directive TORRENT_NO_DEPRECATE and TORRENT_DEBUG(not sure if this is important)
- compiled client project
- it started asking for other dependent dlls like torrent-rasterbar, boost, openssl , added them to runtime directory.
- I dont see linking issue.

On other hand if I run `client.exe base_text.txt` I get ```Exception thrown at 0x00007FF93B2D1314 (vcruntime140d.dll) in client.exe: 0xC0000005: Access violation writing location 0x00007FF7534E7400.``` I dont think this issue is related but I will check further.
